### PR TITLE
[Feature][Connector]add AssertSink connector

### PR DESCRIPTION
**Pull request by lhyundeadsoul**
## Purpose of this pull request
 import an AssertSink plugin, this plugin is similier with ConsoleSink plugin, but will not print data in console, it will validate the data with some rule, e.g: data rows, data type, data value...

https://github.com/apache/incubator-seatunnel/issues/1912

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

**Comment by lhyundeadsoul** [comment]
@ruanwenjun https://github.com/apache/incubator-seatunnel/issues/1912 has been done. Can you help me to review the code ? Thx!

**Comment by lhyundeadsoul** [comment]
AssertSink config is at` examples/fake_to_console.conf:56` ,uncomment and it can be used by `org.apache.seatunnel.example.flink.LocalFlinkExample`

UT is here:
org.apache.seatunnel.flink.assertion.AssertExecutorTest
org.apache.seatunnel.flink.assertion.rule.AssertRuleParserTest

File: seatunnel-connectors/seatunnel-connectors-flink/pom.xml [hunk 1]
```diff
@@ -43,7 +43,7 @@
         <module>seatunnel-connector-flink-doris</module>
         <module>seatunnel-connector-flink-influxdb</module>
         <module>seatunnel-connector-flink-clickhouse</module>
-        <module>seatunnel-connector-flink-http</module>
+        <module>seatunnel-connector-flink-assert</module>
```
Comment (ruanwenjun) [inline seatunnel-connectors/seatunnel-connectors-flink/pom.xml:47]: Please add a new module, rather than change http to assert

File: docs/sidebars.js [hunk 2]
```diff
@@ -128,6 +128,9 @@ const sidebars = {
         'transform/sql',
         'transform/split',
         'transform/json',
+        'transform/replace',
+        'transform/udf',
+        'transform/uuid',
```
Comment (ruanwenjun) [inline docs/sidebars.js]: In my knowledge, the sidebars will be auto generated(I am not sure), but you don't need to do this change in this PR.

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    //User defined rules used to assert illegal data
+    private List<AssertFieldRule> assertFieldRules;
+    private static final String RULES = "rules";
+    private Config config;
+    private List<? extends Config> configList;
+
+    @Override
+    public void outputBatch(FlinkEnvironment env, DataSet<Row> inDataSet) {
+        inDataSet.map(row -> {
+            System.out.println(row);
```
Comment (ruanwenjun) [inline seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java]:
```suggestion
            
```

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    private static final AssertExecutor ASSERT_EXECUTOR = new AssertExecutor();
+    //User defined rules used to assert illegal data
+    private List<AssertFieldRule> assertFieldRules;
+    private static final String RULES = "rules";
+    private Config config;
+    private List<? extends Config> configList;
+
+    @Override
+    public void outputBatch(FlinkEnvironment env, DataSet<Row> inDataSet) {
+        inDataSet.map(row -> {
```
Comment (ruanwenjun) [inline seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java:55]: It's better to use reduce here.

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                        throw new IllegalStateException("row :" + row + " fail rule: " + failRule);
+                    });
+            return null;
+        });
+    }
+
+    @Override
+    public void outputStream(FlinkEnvironment env, DataStream<Row> dataStream) {
+        dataStream.map(row -> {
+            System.out.println(row);
```
Comment (ruanwenjun) [inline seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java]:
```suggestion
        dataStream.reduce(row -> {

```

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/test/java/org/apache/seatunnel/flink/assertion/AssertExecutorTest.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
+package org.apache.seatunnel.flink.assertion;
```
Comment (ruanwenjun) [inline seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/test/java/org/apache/seatunnel/flink/assertion/AssertExecutorTest.java:18]: Please add license header

File: seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/core/flink/config/FlinkRunMode.java [hunk 7]
```diff
@@ -35,4 +35,8 @@ public enum FlinkRunMode {
     public String getMode() {
         return mode;
     }
+
+    public String toString() {
+        return mode;
```
Comment (ruanwenjun) [inline seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/core/flink/config/FlinkRunMode.java]:
```suggestion
        return "FlinkRunMode{" +
            "mode='" + mode + '\'' +
            '}';
```

File: seatunnel-examples/seatunnel-flink-examples/src/main/resources/examples/fake_to_console.conf [hunk 8]
```diff
@@ -48,6 +48,44 @@ transform {
 sink {
   ConsoleSink {}
 
+  #AssertSink {
```
Comment (ruanwenjun) [inline seatunnel-examples/seatunnel-flink-examples/src/main/resources/examples/fake_to_console.conf]: Please revert this change, and add doc for this plugin.

**Review by ruanwenjun — CHANGES_REQUESTED** [review]
Please add this plugin in `seatunnel-connectors-flink-dist/pom.xml`, otherwise the plugin will not in distribution.

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java (hunk 4 above)
Comment (lhyundeadsoul) [reply seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java:55]: Thanks for your suggestion. 
But Reduce function need to process two row data in one place, that looks a little bit strange.

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java (hunk 5 above)
Comment (lhyundeadsoul) [reply seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java]: I didn't want to `system.out` here either at first, but I found that you wouldn't know the progress of code if there is nothing output here when all data is valid. Maybe we can use LOGGER here?

There is no reduce function in dataStream API. If we want to use reduce function, we must keyBy() first.

File: seatunnel-examples/seatunnel-flink-examples/src/main/resources/examples/fake_to_console.conf (hunk 8 above)
Comment (lhyundeadsoul) [reply seatunnel-examples/seatunnel-flink-examples/src/main/resources/examples/fake_to_console.conf]: Can you tell me where can add doc for this plugin?

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java (hunk 4 above)
Comment (ruanwenjun) [reply seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java:55]: Why two row data? In my knowledge, reduce is like map, but it just doesn't need to return data. You don't need to get a result in inDataSet, so you return null here. You can use reduce, then you don't need to return null.

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java (hunk 5 above)
Comment (ruanwenjun) [reply seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java]: Use `System.out.println` is not reliable, it will just print the data into console. If you want to log the data here you can use LOGGER, but I think we don't need to log data here, if there is no error, means the validation is pass.

File: seatunnel-examples/seatunnel-flink-examples/src/main/resources/examples/fake_to_console.conf (hunk 8 above)
Comment (ruanwenjun) [reply seatunnel-examples/seatunnel-flink-examples/src/main/resources/examples/fake_to_console.conf]: You can add log under: https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector/sink

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java (hunk 4 above)
Comment (lhyundeadsoul) [reply seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java:55]: > 
Here is reduce Defination in `org.apache.flink.api.common.functions.ReduceFunction`
`    T reduce(T value1, T value2) throws Exception;`


code will be like
`

    @Override
    public void outputBatch(FlinkEnvironment env, DataSet<Row> inDataSet) {
        inDataSet.reduce((row1, row2) -> {
            ASSERT_EXECUTOR
                    .fail(row1, assertFieldRules)
                    .ifPresent(failRule -> {
                        throw new IllegalStateException("row :" + row1 + " fail rule: " + failRule);
                    });
            ASSERT_EXECUTOR
                    .fail(row2, assertFieldRules)
                    .ifPresent(failRule -> {
                        throw new IllegalStateException("row :" + row2 + " fail rule: " + failRule);
                    });
            return null;
        });
    }
`

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java (hunk 5 above)
Comment (lhyundeadsoul) [reply seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java]: OK, do not log here.

Please take a look this:
There is no reduce function in dataStream API. If we want to use reduce function, we must keyBy() first.

**Comment by lhyundeadsoul** [comment]
There is so many [ImportOrder] checkstyle error.  All of them looks tricky. Is there any approach to  correct my code import order automatically? @ruanwenjun

File: seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java (hunk 4 above)
Comment (lhyundeadsoul) [reply seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java]: code will be like
ima

Comment (ruanwenjun) [reply seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java:55]: Ok, this is not important thing, I close this conversation.

**Comment by ruanwenjun** [comment]
> There is so many [ImportOrder] checkstyle error. All of them looks tricky. Is there any approach to correct my code import order automatically? @ruanwenjun

You need to import `tools/checkstyle/checkstyle.xml` into your IDE.

**Comment by lhyundeadsoul** [comment]
>

File: docs/en/connector/sink/Assert.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+Rule definition of user's available data.  Each rule represents one field validation.
+
+### field_name
+
+field name（string）
+
+### field_type
+
+field type (string),  e.g. `short/int/long/string/boolean`
```
Comment (ruanwenjun) [inline docs/en/connector/sink/Assert.md]: Please give the complete supported field types in the doc, don't just say 
e.g. `short/int/long/string/boolean`, if I see this, I will believe that the assert plugin only support these 5 types.

File: docs/en/connector/sink/Assert.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+### rule_type
+
+The following rules are supported for now
+`
+NOT_NULL,  
+MIN,  
+MAX,  
+MIN_LENGTH,  
+MAX_LENGTH
```
Comment (ruanwenjun) [inline docs/en/connector/sink/Assert.md:54]: Please clarify the meaning of each rule type, and if we break the rule, what will happen.

**Review by ruanwenjun — CHANGES_REQUESTED** [review]

File: seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/batch/FlinkBatchExecution.java [hunk 11]
```diff
@@ -118,6 +118,6 @@ public Config getConfig() {
     }
 
     private boolean whetherExecute(List<FlinkBatchSink> sinks) {
-        return sinks.stream().anyMatch(s -> !"ConsoleSink".equals(s.getPluginName()));
+        return sinks.stream().anyMatch(s -> !"ConsoleSink".equals(s.getPluginName()) && !"AssertSink".equals(s.getPluginName()));
```
Comment (ruanwenjun) [inline seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/batch/FlinkBatchExecution.java]: In my knowledge, you cannot add this check here, this will not call execute method, so your flink job will not be submitted.

**Review by ruanwenjun — CHANGES_REQUESTED** [review]

**Review by ruanwenjun — APPROVED** [review]
LGTM, you can add IT to test this plugin.

**Comment by lhyundeadsoul** [comment]
> LGTM, you can add IT to test this plugin.

Thanks very much for reviewing. Hope to help community develop more efficiently!
